### PR TITLE
Add KeyDecodingStrategy to SQLRow Decode

### PR DESCRIPTION
**Pull request by rnantes**
Adds a key decoding strategy option for decoding `SQLRow`s similar to `JSONDecoder.keyDecodingStrategy`. This is convenient for decoding `SQLRow`s because often table column names are snake_cased whereas codable objects are camelCased.

### CODE REVIEW

**Comment by rnantes** [comment]
Yeah I like how that sounds following closer to `JSONDecoder`. I was also looking into the best way to cache the decoded keys, because as of right now for example if you want to decode 1000 rows and have `keyDecodingStrategy.convertToSnakeCase` set it converts each key to snake case 1000 times.

File: Sources/SQLKit/SQLRow.swift [hunk 1]
```diff
@@ -7,10 +7,19 @@ public protocol SQLRow {
 }
 
 extension SQLRow {
-    public func decode<D>(model type: D.Type, prefix: String? = nil) throws -> D
+    public func decode<D>(model type: D.Type, keyPrefix: String? = nil, keyDecodingStrategy: SQLRowDecoder.KeyDecodingStrategy = .useDefaultKeys) throws -> D
```
Comment (MrLotU) [inline Sources/SQLKit/SQLRow.swift]: Out of pure curiosity, why rename `prefix` to `keyPrefix`?

File: Tests/SQLKitTests/SQLKitTests.swift [hunk 2]
```diff
@@ -348,33 +348,59 @@ CREATE TABLE `planets`(`id` BIGINT, `name` TEXT, `diameter` INTEGER, `galaxy_nam
...
+            let foo = try row.decode(model: Foo.self, keyDecodingStrategy: .convertFromSnakeCase)
+            XCTAssertEqual(foo.foo, 42)
+            XCTAssertEqual(foo.bar, nil)
+            XCTAssertEqual(foo.baz, "vapor")
+            XCTAssertEqual(foo.waldoFred, 2015)
+        } catch {
+            XCTFail("Could not decode row with keyDecodingStrategy \(error)")
         }
     }
 }
```
Comment (MrLotU) [inline Tests/SQLKitTests/SQLKitTests.swift]: Would be great to also have a test with the custom key decoding strategy 😄

Comment (rnantes) [reply Tests/SQLKitTests/SQLKitTests.swift]: I've added a test for this in the latest commit

File: Sources/SQLKit/SQLRow.swift (hunk 1 above)
Comment (rnantes) [reply Sources/SQLKit/SQLRow.swift]: I thought this was more consistent for naming, but I can change it back.

File: Tests/SQLKitTests/SQLKitTests.swift (hunk 2 above)
Comment (rnantes) [reply Tests/SQLKitTests/SQLKitTests.swift]: > Would be great to also have a test with the custom key decoding strategy 😄

I've added a test for this in the latest change.

File: Sources/SQLKit/SQLRowDecoder.swift [hunk 3]
```diff
@@ -1,8 +1,28 @@
-struct SQLRowDecoder {
-    func decode<T>(_ type: T.Type, from row: SQLRow, prefix: String? = nil) throws -> T
+public struct SQLRowDecoder {
+    public var keyPrefix: String? = nil
+    public var keyDecodingStrategy: KeyDecodingStrategy = .useDefaultKeys
```
Comment (tanner0101) [inline Sources/SQLKit/SQLRowDecoder.swift]: Vapor style is to create an explicit `init` method for these (even if `internal`).

File: Sources/SQLKit/SQLRowDecoder.swift [hunk 4]
```diff
@@ -1,8 +1,28 @@
...
+    func decode<T>(_ type: T.Type, from row: SQLRow) throws -> T
         where T: Decodable
     {
-        return try T.init(from: _Decoder(prefix: prefix, row: row))
+        return try T.init(from: _Decoder(row: row, options: options))
+    }
+
+    public enum KeyDecodingStrategy {
+        case useDefaultKeys
+        case convertFromSnakeCase
```
Comment (tanner0101) [inline Sources/SQLKit/SQLRowDecoder.swift]: Some comments here would be nice. I assume this means "convert from snake case to camel case", right?

File: Sources/SQLKit/SQLRowDecoder.swift [hunk 5]
```diff
@@ -82,11 +124,75 @@ struct SQLRowDecoder {
...
         }
 
         func superDecoder(forKey key: Key) throws -> Decoder {
             throw _Error.nesting
         }
     }
 }
+
+fileprivate extension String {
+    /// Returns a new string in snake cased format
```
Comment (tanner0101) [inline Sources/SQLKit/SQLRowDecoder.swift]: Where's this from, Foundation.JSONDecoder? An attribute would be nice.

File: Sources/SQLKit/SQLRow.swift (hunk 1 above)
Comment (tanner0101) [reply Sources/SQLKit/SQLRow.swift]: Since we're in RC phase now we should keep this as `prefix`.

File: Sources/SQLKit/SQLRowDecoder.swift [hunk 6]
```diff
@@ -1,8 +1,28 @@
...
+
+    fileprivate struct _Options {
+        let keyPrefix: String?
+        let keyDecodingStrategy: KeyDecodingStrategy
+    }
+
+    /// The options set on the top-level decoder.
+    fileprivate var options: _Options {
+        return _Options(keyPrefix: keyPrefix,
+                        keyDecodingStrategy: keyDecodingStrategy)
```
Comment (tanner0101) [inline Sources/SQLKit/SQLRowDecoder.swift]: Vapor style is all indents all params on new lines with 1 tab width. (You can see it elsewhere in this repo)

**Pull request merged**